File: packages/tactical-engine/src/targeting/resolve-cone.ts
import type { Coordinate, EdgeObstacle, Obstacle } from "@limiarmap/shared-contracts";
import { blocksEffect } from "../validation/obstacle-rules";
import { canEffectReachAoE } from "./aoe-filter";

/**
 * Resolves a cone-shaped AoE expanding from `origin` toward `anchor`.
 *
 * Phase 11: after computing the geometric footprint, filters each candidate
 * cell through `canEffectReachAoE` from the cone's origin. Cells that are
 * geometrically inside the cone but separated from the origin by a blocking
 * cell or edge obstacle are excluded (shadow zone).
 */
export function resolveCone(
  origin: Coordinate,
  anchor: Coordinate,
  size: number,
  obstacles: Obstacle[],
  edgeObstacles: EdgeObstacle[] = []
): Coordinate[] {
  const result: Coordinate[] = [];
  const deltaX = Math.sign(anchor.x - origin.x);
  const deltaY = Math.sign(anchor.y - origin.y);

  for (let step = 1; step <= size; step += 1) {
    for (let spread = -step + 1; spread <= step - 1; spread += 1) {
      const coordinate =
        Math.abs(deltaX) >= Math.abs(deltaY)
          ? { x: origin.x + deltaX * step, y: origin.y + spread }
          : { x: origin.x + spread, y: origin.y + deltaY * step };

      if (
        !blocksEffect(obstacles, coordinate) &&
        canEffectReachAoE(origin, coordinate, obstacles, edgeObstacles)
      ) {
        result.push(coordinate);
      }
    }
  }

  return result;
}

File: packages/tactical-engine-py/tactical_engine/targeting/resolve_cone.py
from shared_contracts import Coordinate, Obstacle
from ..validation.obstacle_rules import blocks_targeting


def resolve_cone(
    origin: Coordinate,
    anchor: Coordinate,
    size: int,
    obstacles: list[Obstacle],
) -> list[Coordinate]:
    result: list[Coordinate] = []
    delta_x = _sign(anchor.x - origin.x)
    delta_y = _sign(anchor.y - origin.y)

    for step in range(1, size + 1):
        for spread in range(-step + 1, step):
            if abs(delta_x) >= abs(delta_y):
                coordinate = Coordinate(x=origin.x + delta_x * step, y=origin.y + spread)
            else:
                coordinate = Coordinate(x=origin.x + spread, y=origin.y + delta_y * step)
            if not blocks_targeting(obstacles, coordinate):
                result.append(coordinate)

    return result


def _sign(v: int) -> int:
    if v > 0:
        return 1
    if v < 0:
        return -1
    return 0
